services/kvstore/src/main/java/tech/pegasys/pantheon/services/kvstore/KeyValueStorageTransactionTransitionValidatorDecorator.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright 2019 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.pantheon.services.kvstore;
+
+import static com.google.common.base.Preconditions.checkState;
+
+import tech.pegasys.pantheon.plugin.services.exception.StorageException;
+import tech.pegasys.pantheon.plugin.services.storage.KeyValueStorageTransaction;
+
+public class KeyValueStorageTransactionTransitionValidatorDecorator
+    implements KeyValueStorageTransaction {
```

Comment:
I'm not a fan of this decorator. We either want to check whether a transaction has been closed or not. In the implementations of the transactions that would only add a line of code (maybe 3 :-) if you do not call the checkState method).

Reply:
🤷‍♂ There's code reuse (the decorator is used twice, with potential for more with other db plugins) and single responsibility. What's not to like?

Reply:
Performance does suffer, there is that extra class I need to look at to find out what is happening, it only safes us a few lines of code for the cost of an extra Class and two extra method calls.